oshdb-api/src/main/java/org/heigit/ohsome/oshdb/api/mapreducer/MapReducer.java

```diff
@@ -754,20 +755,41 @@ protected MapReducer(MapReducer<?> obj) {
     // apply geometry filter as first map function
     final List<MapFunction> remainingMappers = List.copyOf(ret.mappers);
     ret.mappers.clear();
-    if (ret.forClass.equals(OSMContribution.class)) {
-      ret = ret.filter(x -> {
-        OSMContribution c = (OSMContribution) x;
-        return f.applyOSMGeometry(c.getEntityBefore(), c::getGeometryBefore)
-          || f.applyOSMGeometry(c.getEntityAfter(), c::getGeometryAfter);
-      });
-    } else if (ret.forClass.equals(OSMEntitySnapshot.class)) {
-      ret = ret.filter(x -> {
-        OSMEntitySnapshot s = (OSMEntitySnapshot) x;
-        return f.applyOSMGeometry(s.getEntity(), s::getGeometry);
-      });
+    if (this.grouping == Grouping.BY_ID) {
+      if (ret.forClass.equals(OSMContribution.class)) {
+        @SuppressWarnings("unchecked") MapReducer<X> result = (MapReducer<X>)
+            ret.map(x -> (Collection<OSMContribution>) x)
+                .map(contributions -> Collections2.filter(contributions, c -> c != null
```

Comment:
todo: check if it's save to assume that this is still a `List` (not just a generic `Collection` which doesn't implement stuff like `get(index)`